src/vlrs.jl

```diff
@@ -124,33 +124,69 @@ The registered record IDs associated to VLRs of this record type `TData`
 """
 official_record_ids(::Type{TData}) where TData = error("Official record IDs not set for VLRs with data type $(TData)")
 
+"""
+    $(TYPEDSIGNATURES)
+
+Get the data type associated with a particular `user_id` and `record_id`. 
+This is used to automatically parse VLR data types on reading
+"""
+function data_type_from_ids(user_id::String, record_id::Integer)
+    registered_vlr_types = get_all_vlr_types()
+
+    for type ∈ registered_vlr_types
+        if (user_id == official_user_id(type)) && (record_id ∈ official_record_ids(type))
+            return type
+        end
+    end
+
+    error("Can't find VLR data type to parse for user ID $(user_id) and record ID $(record_id)")
+end
+
+function get_all_vlr_types()
+    # eval to make sure we get the methods in the global scope - use both the user and record IDs methods to check for types
+    ms = @eval vcat(methods(LASDatasets.official_user_id), methods(LASDatasets.official_record_ids))
+    types = Set{DataType}()
+    for m ∈ ms
+        sig = m.sig
+        if !(sig isa UnionAll)
+            # need to convert from a TypeVar to a DataType using eval
+            push!(types, eval(sig.parameters[2].parameters[1]))
+        end
```

Comment:
![image](https://github.com/user-attachments/assets/40714035-7364-4437-8de6-d4d6f15bec2f)
method signatures are things I try to avoid... But maybe can simplify by removing the Type in the method args and can  just push the sig.parameters[2] in?

Reply:
oh wait the Val things!!!
![image](https://github.com/user-attachments/assets/5c5f2f20-f5ce-4b6f-8bff-0092180181cd)

Reply:
The main reason I wanted to avoid using `Val` was so we would have less functions that needed to be compiled into a system image which could increase the size a bit. As a test I did a couple of modules (one with the `Val` stuff, one without) and compiled them using PackageCompiler.jl to compare the sizes:

With Val: sys image size 197.1MB
```julia
module MethodsTest

function test end

macro register_type(type, user, id)
    return quote
        MethodsTest.test(::Val{Symbol($(esc(user)))}, ::Val{$(esc(id))}) = $(esc(type))
    end
end

types = (String, Float16, Float32, Float64, UInt8, UInt16, UInt32, UInt64, Int8, Int16, Int32, Int64)
user_ids = ntuple(i -> "$(i)", length(types))

for i ∈ eachindex(types)
    for id ∈ 1:1000
        @register_type types[i] user_ids[i] id
    end
end

export test

end # module MethodsTest
```

Without Val: sysimage size: 186.5MB
```julia
module MoreMethods

function official_user_ids end
function official_record_ids end

macro register_type(type, user_id, record_ids)
    return quote
        MoreMethods.official_user_ids(::Type{$(esc(type))}) = $(user_id)
        MoreMethods.official_record_ids(::Type{$(esc(type))}) = record_ids
    end
end

types = (String, Float16, Float32, Float64, UInt8, UInt16, UInt32, UInt64, Int8, Int16, Int32, Int64)
user_ids = ntuple(i -> "$(i)", length(types))

for i ∈ eachindex(types)
    @register_type types[i] user_ids[i] 1:1000
end

end # module MoreMethods
```

I'd anticipate this being more an issue with the more VLR types you want but might depend?